command/agent/cache/handler.go

```diff
@@ -0,0 +1,155 @@
+package cache
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/hashicorp/errwrap"
+	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/helper/consts"
+	vaulthttp "github.com/hashicorp/vault/http"
+	"github.com/hashicorp/vault/logical"
+)
+
+func Handler(ctx context.Context, logger hclog.Logger, proxier Proxier, useAutoAuthToken bool, client *api.Client) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		logger.Info("received request", "path", r.URL.Path, "method", r.Method)
+
+		token := r.Header.Get(consts.AuthHeaderName)
+		if token == "" && useAutoAuthToken {
+			logger.Debug("using auto auth token")
+			token = client.Token()
+		}
+
+		// Parse and reset body.
+		reqBody, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			logger.Error("failed to read request body")
+			respondError(w, http.StatusInternalServerError, errors.New("failed to read request body"))
+		}
+		if r.Body != nil {
+			r.Body.Close()
```

Comment:
Same comments here -- move the check up, and why is the next line there?

Reply:
Yep, we can move the nil check before the read, but same idea applies from the other comment as to why we're re-setting r.Body. Each of the layers might need to read the request, so we do that here before we send it to the first layer on `proxier.Send` so that it can simply read the request body using `SendRequest.RequestBody`. We reset the buffer using ioutil.NopCloser because when we get to the proxy layer the client uses the actual `r.Body`. We _could_ re-set it right before proxying, but we might as well reset it here.